### PR TITLE
PHPLIB-1236 Implement Multi-Doc Benchmarks

### PR DESCRIPTION
**Pull request by GromNaN**
Fix [PHPLIB-1236](https://jira.mongodb.org/browse/PHPLIB-1236)

https://github.com/mongodb/specifications/blob/master/source/benchmarking/benchmarking.rst#multi-doc-benchmarks

- [x] Find many and empty the cursor
- [x] Small doc bulk insert
- [x] Large doc bulk insert
- [x] GridFS upload
- [x] GridFS download

```
\MongoDB\Benchmark\DriverBench\MultiDocBench

    benchFindMany # Driver default typemap..I2 - Mo133.957ms (±0.20%)
    benchFindMany # Raw BSON................I2 - Mo17.109ms (±0.86%)
    benchBulkInsert # Small doc.............I2 - Mo53.328ms (±0.25%)
    benchBulkInsert # Small BSON doc........I2 - Mo53.180ms (±0.36%)
    benchBulkInsert # Large doc.............I2 - Mo242.104ms (±2.10%)
    benchBulkInsert # Large BSON doc........I2 - Mo239.291ms (±0.63%)

\MongoDB\Benchmark\DriverBench\GridFSBench

    benchUpload.............................I2 - Mo31.448ms (±0.47%)
    benchDownload...........................I2 - Mo67.099ms (±11.71%)

Subjects: 4, Assertions: 0, Failures: 0, Errors: 0
```

### CODE REVIEW

File: benchmark/Fixtures/Data.php [hunk 1]
```diff
@@ -17,4 +21,18 @@ public static function readJsonFile(string $path): array
...
+
+    /**
+     * Generates an in-memory stream of the given size.
+     *
+     * @return resource
+     */
+    public static function getStream(int $size)
+    {
+        $stream = fopen('php://memory', 'w+');
+        fwrite($stream, str_repeat("\0", $size));
```
Comment (GromNaN) [inline benchmark/Fixtures/Data.php:33]: > The dataset, designated GRIDFS_LARGE (disk file 'gridfs_large.bin'), consists of a single file containing about 50 MB of random data.

I don't need to commit a 50MB file full of NULL characters from the spec. I can generate it on-demand in-memory.

File: benchmark/Utils.php [hunk 2]
```diff
@@ -4,41 +4,43 @@
...
+    private static ?Database $database;
     private static ?Collection $collection;
 
     public static function getClient(): Client
     {
-        return self::$client ??= self::createClient();
+        return self::$client ??= new Client(self::getUri());
+    }
+
+    public static function getDatabase(): Database
```
Comment (GromNaN) [inline benchmark/Utils.php:22]: Refactored to give access to an instance of each library objects, and cache them.

File: phpbench.json.dist [hunk 3]
```diff
@@ -5,6 +5,7 @@
     "runner.bootstrap": "vendor/autoload.php",
     "runner.file_pattern": "*Bench.php",
     "runner.path": "benchmark",
+    "runner.php_config": { "memory_limit": "1G" },
```
Comment (GromNaN) [inline phpbench.json.dist:8]: 50MB file takes a lot more than expected to be downloaded in-memory.

**Review by alcaeus — APPROVED** [review]

File: phpbench.json.dist (hunk 3 above)
Comment (GromNaN) [reply phpbench.json.dist:8]: I need to profile why this is taking so much memory.

Comment (GromNaN) [reply phpbench.json.dist:8]: For reference, it's the `substr` that uses the most memory.
https://github.com/mongodb/mongo-php-library/blob/ec6c431acc87ced5e85a163b92e76a8d8f69e000/src/GridFS/ReadableStream.php#L166-L174 
image

**Pull request merged**